File: src/merge_sort.rs
//fix
pub fn  merge(arr: &mut Vec<i32>,ini: usize,mitad: usize,fin: usize){
    //create copies (left and right)
    let mut left: Vec<i32>=Vec::new();
    let mut right: Vec<i32>=Vec::new();
    for i in ini..mitad{left.push(arr[i]);}
    for j in mitad..fin{right.push(arr[j]);}
    let (mut a,mut b,mut k)=(0,0,ini);
    while a<left.len() && b<right.len(){
        if left[a]<right[b]{
            arr[k]=left[a];
            a+=1;
        }else if left[a]>=right[b]{
            arr[k]=left[b];
            b+=1; 
        }
        k+=1;
    }
    while a<left.len(){
        arr[k]=left[a];
        k+=1;
        a+=1;
    }
    while b<right.len(){
        arr[k]=right[b];
        k+=1;
        b+=1;
    }
}
pub fn merge_sort_algorithm(arr: &mut Vec<i32>,start: usize,end: usize){
    if start+1<end{
        let mid =start+(end-start)/2;
        merge_sort_algorithm(arr,start,mid);
        merge_sort_algorithm(arr,mid+1,end);            
        merge(arr,start,mid,end); //last task for every recursion!
    }
}
pub fn merge_sort(arr: &mut Vec<i32>){
    merge_sort_algorithm(arr, 0, arr.len());
}

File: java-codes/MergeSort.java
public class MergeSort {
    public static void merge_sort(int[] arr,int start,int end){
        if(start < end){
            int mid=start+(end-start)/2;
            merge_sort(arr, start, mid);
            merge_sort(arr, mid+1, end);
            merge(arr,start,end);
        }
    }
    public static void merge(int[] arr,int start,int end){
        int mid=start+(end-start)/2;
        int left[]=new int[mid-start+1];
        int rigth[]=new int[end-mid];
        for(int i=start;i<=mid;i++){
            left[i-start]=arr[i];
        }
        for(int j=mid;j<end;j++){
            rigth[j-mid]=arr[j+1];
        }
        int i=0,j=0; 
        while(i<left.length && j<rigth.length){
            if(left[i]<rigth[j]){
                arr[start]=left[i];
                i++;
            }else{
                arr[start]=rigth[j];
                j++;
            }
            start++;
        }
        while(i<left.length){
            arr[start++]=left[i++];
        }
        while(j<rigth.length){
            arr[start++]=rigth[j++];
        }
    }
    public static void main(String[] args) {
        int[] array={7,5,4,1,8,2,3,10,3,21,5,54,71,103,-1,-98,-54};
        int n=array.length;
        merge_sort(array,0,n-1);
        for(int i=0;i<n;i++){
            System.out.print(array[i]+" ");
        }
    }
}
